### PR TITLE
udp: add UV_UDP_MMSG_FREE recv_cb flag

### PR DESCRIPTION
**Pull request by squeek502**
Contributes towards #2822. From https://github.com/libuv/libuv/issues/2822#issuecomment-624512182:

> When using `UV_UDP_RECVMMSG`, `recv_cb` gets called once for each message with the `UV_UDP_MMSG_CHUNK` flag set, and then [once with `nread=0`, `addr=NULL`, and without the `UV_UDP_MMSG_CHUNK` flag set to free the buffer](https://github.com/libuv/libuv/blob/e7ebae26247d2fee0a04547eb7f9aa8f78d4a642/src/unix/udp.c#L234-L236).
> 
> For binding this in a garbage collected language, passing that last callback on seems somewhat strange, as the garbage collected language shouldn't care about it, and it would cause a discrepancy in the number of callbacks called when using `mmsg` and when not using it. That is, using the test added in #2818 as an example, where using `mmsg` leads to 2 allocs for 8 messages:
> 
> * When not using `UV_UDP_RECVMMSG`, `recv_cb` would be called 8 times (once for each message)
> * When using `UV_UDP_RECVMMSG`, `recv_cb` would be called 10 times (once for each message, and then 2 for freeing the mmsg buffers)
> 
> Not passing these 'free-only' callbacks on would be no problem except that there doesn't seem to be a way to distinguish these 'free-only' callbacks from things like [callbacks from `EAGAIN` or `EWOULDBLOCK`](https://github.com/libuv/libuv/blob/e7ebae26247d2fee0a04547eb7f9aa8f78d4a642/src/unix/udp.c#L292-L293), which doesn't seem ideal. But, again, I might be thinking about this wrong, so any feedback is appreciated.

For an example of how this would be used, see https://github.com/squeek502/libuv/commit/94de1d6af2a716fd832fe3fec4c9d5f46d781547 which alters the test added in #2818 to use `UV_UDP_MMSG_FREE`.

### CODE REVIEW

File: docs/src/udp.rst [hunk 1]
```diff
@@ -80,8 +86,10 @@ Data types
     When using :man:`recvmmsg(2)`, chunks will have the `UV_UDP_MMSG_CHUNK` flag set,
     those must not be freed. There will be a final callback with `nread` set to 0,
     `addr` set to NULL and the buffer pointing at the initially allocated data with
-    the `UV_UDP_MMSG_CHUNK` flag cleared. This is a good chance for the callee to
-    free the provided buffer.
+    the `UV_UDP_MMSG_CHUNK` flag cleared and the `UV_UDP_MMSG_FREE` flag set.
+    This is a good chance for the callee to free the provided buffer.
```
Comment (bnoordhuis) [inline docs/src/udp.rst:90]:
```suggestion
    The callee can now safely free the provided buffer.
```
Or something along those lines? "A good chance" sounds somewhat inconclusive.

**Comment by squeek502** [comment]
> It needs a test though. :-)

Would be able to easily test it once https://github.com/libuv/libuv/pull/2818 is merged.

**Comment by squeek502** [comment]
Test added, this should be good to go.

**Comment by vtjnash** [comment]
CI: https://ci.nodejs.org/job/libuv-test-commit/1964/

**Comment by vtjnash** [comment]
@bnoordhuis should I consider that LGTM a Reviewed-By, or did you want to look again now that there's a test? I'm partly just a rubber stamp, since you said you think it is a good change.

File: include/uv.h [hunk 2]
```diff
@@ -613,6 +613,12 @@ enum uv_udp_flags {
    * must not be freed by the recv_cb callback.
    */
   UV_UDP_MMSG_CHUNK = 8,
+  /*
+   * Indicates that the buffer provided has been fully utilized by recvmmsg and that
```
Comment (cjihrig) [inline include/uv.h:617]: Can you wrap lines at 80 characters please.

**Comment by squeek502** [comment]
Any reason this didn't make it into 1.39.0?

**Comment by cjihrig** [comment]
Sorry for the delay. Landed in https://github.com/libuv/libuv/commit/ea17e1cffb7bb379191c0efd0e264d7bfdafd147!

(I added @bnoordhuis as a reviewer in the commit metadata due to the "LGTM." I hope that was the right thing to do.)

**Comment by squeek502** [comment]
No worries! I think [this line](https://github.com/libuv/libuv/commit/ea17e1cffb7bb379191c0efd0e264d7bfdafd147#diff-88efe1766acb2d2af3a27c1a33dc0510R92) needs to be updated to 1.40.0 (presumably). Not totally sure how/when you usually handle that, though.

**Comment by cjihrig** [comment]
🤦 yea, it needs to be updated.

**Pull request closed without merging**